File: modules/Scripting/modules/RmlUi/src/Style.cpp
#include "Style.hpp"
#include "RmlUi/Core/Box.h"
#include "RmlUi/Core/StyleTypes.h"
#include "RmlUi/Core/StyleSheet.h"
#include "RmlUi/Core/StyleSheetSpecification.h"
#include "glm2Rml.hpp"
#include "Rml2glm.hpp"

#include "Sol2HelperMacros.hpp"

using namespace Rml;

namespace {

void registerStyleTypes(sol::table &lua) {
  using namespace Style;

  // clang-format off
#define MAKE_PAIR(Value) _MAKE_PAIR(Display, Value)
  lua.DEFINE_ENUM(Display, {
    MAKE_PAIR(None),
    MAKE_PAIR(Block),
    MAKE_PAIR(Inline),
    MAKE_PAIR(InlineBlock),
    MAKE_PAIR(Flex),
    MAKE_PAIR(Table),
    MAKE_PAIR(TableRow),
    MAKE_PAIR(TableRowGroup),
    MAKE_PAIR(TableColumn),
    MAKE_PAIR(TableColumnGroup),
    MAKE_PAIR(TableCell),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(Position, Value)
  lua.DEFINE_ENUM(Position, {
    MAKE_PAIR(Static),
    MAKE_PAIR(Relative),
    MAKE_PAIR(Absolute),
    MAKE_PAIR(Fixed),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(Float, Value)
  lua.DEFINE_ENUM(Float, {
    MAKE_PAIR(None),
    MAKE_PAIR(Left),
    MAKE_PAIR(Right),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(Clear, Value)
  lua.DEFINE_ENUM(Clear, {
    MAKE_PAIR(None),
    MAKE_PAIR(Left),
    MAKE_PAIR(Right),
    MAKE_PAIR(Both),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(BoxSizing, Value)
  lua.DEFINE_ENUM(BoxSizing, {
    MAKE_PAIR(ContentBox),
    MAKE_PAIR(BorderBox),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(Overflow, Value)
  lua.DEFINE_ENUM(Overflow, {
    MAKE_PAIR(Visible),
    MAKE_PAIR(Hidden),
    MAKE_PAIR(Auto),
    MAKE_PAIR(Scroll),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(Visibility, Value)
  lua.DEFINE_ENUM(Visibility, {
    MAKE_PAIR(Visible),
    MAKE_PAIR(Hidden),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(FontStyle, Value)
  lua.DEFINE_ENUM(FontStyle, {
    MAKE_PAIR(Normal),
    MAKE_PAIR(Italic),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(FontWeight, Value)
  lua.DEFINE_ENUM(FontWeight, {
    MAKE_PAIR(Auto),
    MAKE_PAIR(Normal),
    MAKE_PAIR(Bold),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(TextAlign, Value)
  lua.DEFINE_ENUM(TextAlign, {
    MAKE_PAIR(Left),
    MAKE_PAIR(Right),
    MAKE_PAIR(Center),
    MAKE_PAIR(Justify),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(TextDecoration, Value)
  lua.DEFINE_ENUM(TextDecoration, {
    MAKE_PAIR(None),
    MAKE_PAIR(Underline),
    MAKE_PAIR(Overline),
    MAKE_PAIR(LineThrough),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(TextTransform, Value)
  lua.DEFINE_ENUM(TextTransform, {
    MAKE_PAIR(None),
    MAKE_PAIR(Capitalize),
    MAKE_PAIR(Uppercase),
    MAKE_PAIR(Lowercase),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(WhiteSpace, Value)
  lua.DEFINE_ENUM(WhiteSpace, {
    MAKE_PAIR(Normal),
    MAKE_PAIR(Pre),
    MAKE_PAIR(Nowrap),
    MAKE_PAIR(Prewrap),
    MAKE_PAIR(Preline),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(WordBreak, Value)
  lua.DEFINE_ENUM(WordBreak, {
    MAKE_PAIR(Normal),
    MAKE_PAIR(BreakAll),
    MAKE_PAIR(BreakWord),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(Drag, Value)
  lua.DEFINE_ENUM(Drag, {
    MAKE_PAIR(None),
    MAKE_PAIR(Drag),
    MAKE_PAIR(DragDrop),
    MAKE_PAIR(Block),
    MAKE_PAIR(Clone),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(TabIndex, Value)
  lua.DEFINE_ENUM(TabIndex, {
    MAKE_PAIR(None),
    MAKE_PAIR(Auto),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(Focus, Value)
  lua.DEFINE_ENUM(Focus, {
    MAKE_PAIR(None),
    MAKE_PAIR(Auto),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(OverscrollBehavior, Value)
  lua.DEFINE_ENUM(OverscrollBehavior, {
    MAKE_PAIR(Auto),
    MAKE_PAIR(Contain),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(PointerEvents, Value)
  lua.DEFINE_ENUM(PointerEvents, {
    MAKE_PAIR(None),
    MAKE_PAIR(Auto),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(OriginX, Value)
  lua.DEFINE_ENUM(OriginX, {
    MAKE_PAIR(Left),
    MAKE_PAIR(Center),
    MAKE_PAIR(Right),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(OriginY, Value)
  lua.DEFINE_ENUM(OriginY, {
    MAKE_PAIR(Top),
    MAKE_PAIR(Center),
    MAKE_PAIR(Bottom),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(AlignContent, Value)
  lua.DEFINE_ENUM(AlignContent, {
    MAKE_PAIR(FlexStart),
    MAKE_PAIR(FlexEnd),
    MAKE_PAIR(Center),
    MAKE_PAIR(SpaceBetween),
    MAKE_PAIR(SpaceAround),
    MAKE_PAIR(Stretch),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(AlignItems, Value)
  lua.DEFINE_ENUM(AlignItems, {
    MAKE_PAIR(FlexStart),
    MAKE_PAIR(FlexEnd),
    MAKE_PAIR(Center),
    MAKE_PAIR(Baseline),
    MAKE_PAIR(Stretch),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(AlignSelf, Value)
  lua.DEFINE_ENUM(AlignSelf, {
    MAKE_PAIR(Auto),
    MAKE_PAIR(FlexStart),
    MAKE_PAIR(FlexEnd),
    MAKE_PAIR(Center),
    MAKE_PAIR(Baseline),
    MAKE_PAIR(Stretch),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(FlexDirection, Value)
  lua.DEFINE_ENUM(FlexDirection, {
    MAKE_PAIR(Row),
    MAKE_PAIR(RowReverse),
    MAKE_PAIR(Column),
    MAKE_PAIR(ColumnReverse),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(FlexWrap, Value)
  lua.DEFINE_ENUM(FlexWrap, {
    MAKE_PAIR(Nowrap),
    MAKE_PAIR(Wrap),
    MAKE_PAIR(WrapReverse),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(JustifyContent, Value)
  lua.DEFINE_ENUM(JustifyContent, {
    MAKE_PAIR(FlexStart),
    MAKE_PAIR(FlexEnd),
    MAKE_PAIR(Center),
    MAKE_PAIR(SpaceBetween),
    MAKE_PAIR(SpaceAround),
  });
#undef MAKE_PAIR
  // clang-format on
}

void registerBox(sol::table &lua) {
  // clang-format off
#define MAKE_PAIR(Value) _MAKE_PAIR(BoxArea, Value)
  lua.DEFINE_ENUM(BoxArea, {
    MAKE_PAIR(Margin),
    MAKE_PAIR(Border),
    MAKE_PAIR(Padding),
    MAKE_PAIR(Content),
    MAKE_PAIR(Auto),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(BoxEdge, Value)
  lua.DEFINE_ENUM(BoxEdge, {
    MAKE_PAIR(Top),
    MAKE_PAIR(Right),
    MAKE_PAIR(Bottom),
    MAKE_PAIR(Left),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(BoxDirection, Value)
  lua.DEFINE_ENUM(BoxDirection, {
    MAKE_PAIR(Vertical),
    MAKE_PAIR(Horizontal),
  });
#undef MAKE_PAIR

  lua.DEFINE_USERTYPE(Box,
    sol::call_constructor,
    sol::factories(
      []{ return Box{}; },
      [](const glm::vec2 content){ return Box{to_Rml(content)}; }
    ),
    
    sol::meta_function::equal_to, &Box::operator==,

    "getPosition", sol::overload(
      [](const Box &self) { return to_glm(self.GetPosition()); },
      [](const Box &self, BoxArea area) {
        return to_glm(self.GetPosition(area));
      }
    ),
    "getSize", [](const Box &self, BoxArea area) {
      return to_glm(self.GetSize(area));
    },
    "setContent", [](Box &self, const glm::vec2 content) {
      self.SetContent(to_Rml(content));
    },
    "setEdge", &Box::SetEdge,
    "getEdge", &Box::GetEdge,
    "getCumulativeEdge", &Box::GetCumulativeEdge,
    "getSizeAcross", sol::overload(
      [](const Box &self, BoxDirection direction, BoxArea area) {
        return self.GetSizeAcross(direction, area);
      },
      &Box::GetSizeAcross
    )
  );
  // clang-format on
}
void registerStyleSheet(sol::table &lua) {
  lua.DEFINE_USERTYPE(StyleSheet, sol::no_constructor);
}
void registerStyleSheetSpecification(sol::table &lua) {
  // clang-format off
  lua.DEFINE_USERTYPE(StyleSheetSpecification,
    sol::no_constructor,

    "parsePropertyDeclaration", StyleSheetSpecification::ParsePropertyDeclaration,
  
    "getPropertyId", StyleSheetSpecification::GetPropertyId,
    "getShorthandId", StyleSheetSpecification::GetShorthandId,
    "getPropertyName", StyleSheetSpecification::GetPropertyName,
    "getShorthandName", StyleSheetSpecification::GetShorthandName
  );
  // clang-format on
}

} // namespace

void registerStyle(sol::table &lua) {
  auto m = lua["Style"].get_or_create<sol::table>();
  registerStyleTypes(m);

  registerBox(lua);
  registerStyleSheet(lua);
  registerStyleSheetSpecification(lua);
}

File: modules/Scripting/annotations/RmlUi/Style.lua
---@meta

ui.Style = {}

---@enum ui.Style.Display
ui.Style.Display = {
  None = 0,
  Block = 1,
  Inline = 2,
  InlineBlock = 3,
  Flex = 4,
  Table = 5,
  TableRow = 6,
  TableRowGroup = 7,
  TableColumn = 8,
  TableColumnGroup = 9,
  TableCell = 10,
}

---@enum ui.Style.Position
ui.Style.Position = {
  Static = 0, Relative = 1, Absolute = 2, Fixed = 3,
}

---@enum ui.Style.Float
ui.Style.Float = {
  None = 0, Left = 1, Right = 2,
}

---@enum ui.Style.Clear
ui.Style.Clear = {
  None = 0, Left = 1, Right = 2, Both = 3,
}

---@enum ui.Style.BoxSizing
ui.Style.BoxSizing = {
  ContentBox = 0, BorderBox = 1,
}

---@enum ui.Style.Overflow
ui.Style.Overflow = {
  Visible = 0, Hidden = 1, Auto = 2, Scroll = 3,
}

---@enum ui.Style.Visibility
ui.Style.Visibility = {
  Visible = 0, Hidden = 1,
}

---@enum ui.Style.FontStyle
ui.Style.FontStyle = {
  Normal = 0, Italic = 1,
}

---@enum ui.Style.FontWeight
ui.Style.FontWeight = {
  Auto = 0, Normal = 400, Bold = 700,
}

---@enum ui.Style.TextAlign
ui.Style.TextAlign = {
  Left = 0, Right = 1, Center = 2, Justify = 3,
}

---@enum ui.Style.TextDecoration
ui.Style.TextDecoration = {
  None = 0, Underline = 1, Overline = 2, LineThrough = 3,
}

---@enum ui.Style.TextTransform
ui.Style.TextTransform = {
  None = 0, Capitalize = 1, Uppercase = 2, Lowercase = 3,
}

---@enum ui.Style.WhiteSpace
ui.Style.WhiteSpace = {
  Normal = 0, Pre = 1, Nowrap = 2, Prewrap = 3, Preline = 4,
}

---@enum ui.Style.WordBreak
ui.Style.WordBreak = {
  Normal = 0, BreakAll = 1, BreakWord = 2,
}

---@enum ui.Style.Drag
ui.Style.Drag = {
  None = 0, Drag = 1, DragDrop = 2, Block = 3, Clone = 4,
}

---@enum ui.Style.TabIndex
ui.Style.TabIndex = {
  None = 0, Auto = 1,
}

---@enum ui.Style.Focus
ui.Style.Focus = {
  None = 0, Auto = 1,
}

---@enum ui.Style.OverscrollBehavior
ui.Style.OverscrollBehavior = {
  Auto = 0, Contain = 1,
}

---@enum ui.Style.PointerEvents
ui.Style.PointerEvents = {
  None = 0, Auto = 1,
}

---@enum ui.Style.OriginX
ui.Style.OriginX = {
  Left = 0, Center = 1, Right = 2,
}

---@enum ui.Style.OriginY
ui.Style.OriginY = {
  Top = 0, Center = 1, Bottom = 2,
}

---@enum ui.Style.AlignContent
ui.Style.AlignContent = {
  FlexStart = 0,
  FlexEnd = 1,
  Center = 2,
  SpaceBetween = 3,
  SpaceAround = 4,
  Stretch = 5,
}

---@enum ui.Style.AlignItems
ui.Style.AlignItems = {
  FlexStart = 0,
  FlexEnd = 1,
  Center = 2,
  Baseline = 3,
  Stretch = 4,
}

---@enum ui.Style.AlignSelf
ui.Style.AlignSelf = {
  Auto = 0,
  FlexStart = 1,
  FlexEnd = 2,
  Center = 3,
  Baseline = 4,
  Stretch = 5,
}

---@enum ui.Style.FlexDirection
ui.Style.FlexDirection = {
  Row = 0, RowReverse = 1, Column = 2, ColumnReverse = 3,
}

---@enum ui.Style.FlexWrap
ui.Style.FlexWrap = {
  Nowrap = 0, Wrap = 1, WrapReverse = 2,
}

---@enum ui.Style.JustifyContent
ui.Style.JustifyContent = {
  FlexStart = 0,
  FlexEnd = 1,
  Center = 2,
  SpaceBetween = 3,
  SpaceAround = 4,
}

---@enum ui.BoxArea
ui.BoxArea = {
  MARGIN = 0,
  BORDER = 1,
  PADDING = 2,
  CONTENT = 3,
  NUM_AREAS = 3,
}

---@enum ui.BoxEdge
ui.BoxEdge = {
  TOP = 0,
  RIGHT = 1,
  BOTTOM = 2,
  LEFT = 3,
  NUM_EDGES = 4,
}

---@enum ui.BoxDirection
ui.BoxDirection = {
  VERTICAL = 0, HORIZONTAL = 1,
}

---@class ui.Box
---@overload fun(content?: vec2): ui.Box
ui.Box = {}

---Returns the top-left position of one of the box's areas, relative to the top-left of the border area. This
---means the position of the margin area is likely to be negative.
---@param area? ui.BoxArea # The desired area.
---@return vec2 # The position of the area.
function ui.Box:getPosition(area) end

---Returns the size of one of the box's areas. This will include all inner areas.
---@param area ui.BoxArea # The desired area.
---@return vec2 # The size of the requested area.
---@overload fun(): vec2 # The size of the content area.
function ui.Box:getSize(area) end

---Sets the size of the content area.
---@param content vec2 @ The size of the new content area.
function ui.Box:setContent(content) end

---Sets the size of one of the edges of one of the box's outer areas.
---@param area ui.BoxArea # The area to change.
---@param edge ui.BoxEdge # The area edge to change.
---@param size number # The new size of the area segment.
function ui.Box:setEdge(area, edge, size) end

---Returns the size of one of the area edges.
---@param area ui.BoxArea # The desired area.
---@param edge ui.BoxEdge # The desired edge.
---@return number # The size of the requested area edge.
function ui.Box:getEdge(area, edge) end

---Returns the cumulative size of one edge up to one of the box's areas.
---@param area ui.BoxArea # The area to measure up to (and including). So, MARGIN will return the width of the margin, and PADDING will be the sum of the margin, border and padding.
---@param edge ui.BoxEdge # The desired edge.
---@return number # The cumulative size of the edge.
function ui.Box:getCumulativeEdge(area, edge) end

---Returns the size along a single direction of the given 'area', including all inner areas up-to and including 'area_end'.<br/>
---EXAMPLE: getSizeAcross(HORIZONTAL, BORDER, PADDING) returns the total width of the horizontal borders and paddings.
---@param direction ui.BoxDirection # The desired direction.
---@param area ui.BoxArea # The widest area to include.
---@param areaEnd? ui.BoxArea # The last area to include, anything inside this is excluded.
function ui.Box:getSizeAcross(direction, area, areaEnd) end

---@class ui.StyleSheet
ui.StyleSheet = {}

---@class ui.StyleSheetSpecification
ui.StyleSheetSpecification = {}

---Parses a property declaration, setting any parsed and validated properties on the given dictionary.
---@param dictionary ui.PropertyDictionary # The property dictionary which will hold all declared properties.
---@param propertyName string # The name of the declared property.
---@param propertyValue string # The values the property is being set to.
---@return boolean # True if all properties were parsed successfully, false otherwise.
function ui.StyleSheetSpecification.parsePropertyDeclaration(dictionary, propertyName, propertyValue) end

---@param name string
---@return ui.PropertyId
function ui.StyleSheetSpecification.getPropertyId(name) end

---@param name string
---@return ui.ShorthandId
function ui.StyleSheetSpecification.getShorthandId(name) end

---@param id ui.PropertyId
---@return string
function ui.StyleSheetSpecification.getPropertyName(id) end

---@param id ui.ShorthandId
---@return string
function ui.StyleSheetSpecification.getShorthandName(id) end
